Ignore derived Clone and Debug implementations during dead code analysis

FabianWolff opened the pull request:
This pull request fixes #84647. Derived implementations of `Clone` and `Debug` always trivially read all fields, so "field is never read" dead code warnings are never triggered. Arguably, though, a user most likely will only be interested in whether _their_ code ever reads those fields, which is the behavior I have implemented here.

Note that implementations of `Clone` and `Debug` are only ignored if they are `#[derive(...)]`d; a custom `impl Clone/Debug for ...` will still be analyzed normally (i.e. if a custom `Clone` implementation uses all fields of the struct, this will continue to suppress dead code warnings about unused fields); this seemed like the least intrusive change to me (although it would be easy to change — just drop the `&& [impl_]item.span.in_derive_expansion()` in the if conditions).

The only thing that I am slightly unsure about is that in #84647, @matklad said
> Doesn't seem easy to fix though :(

However, it _was_ pretty straightforward to fix, so did I perhaps overlook something obvious? @matklad, could you weigh in on this?

rust-highfive commented:
r? @lcnr

(rust-highfive has picked a reviewer for you, use r? to override)

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2560635103) [(plain)](https://github.com/rust-lang/rust/commit/c0613f7b1540f8f5a7cf0101cda06785f2f52735/checks/2560635103/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   |
54 |     description: &'static str,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `-D dead-code` implied by `-D warnings`
error: aborting due to previous error

error: could not compile `rustc_feature`

```

</details>

FabianWolff commented:
As an argument in favor of this pull request, it has already found a number of unused fields in the compiler code, breaking the bootstrap process. I have now sprinkled a few `#[allow(dead_code)]` (with comments) throughout the code to make it build again; of course, some of those fields actually should be dropped, but I lack the knowledge about how and where all of those structs are used to make this decision (perhaps instead of dropping the field, it should be used somewhere).

I would have preferred `#[warn(dead_code)]` instead of `#[allow(dead_code)]`, but this does not work with `-D warnings` (it always emits errors, even with `#[warn(...)]`, see #75668).

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2565986911) [(plain)](https://github.com/rust-lang/rust/commit/bb8ca5ab9f65af558601482f628242ceb13b27a0/checks/2565986911/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
.................................................................................................... 9500/11852
.................................................................................................... 9600/11852
.............................................................i......i............................... 9700/11852
.................................................................................................... 9800/11852
.......iiiiiii..iiiiii.i............................................................................ 9900/11852
.................................................................................................... 10100/11852
.................................................................................................... 10200/11852
.................................................................................................... 10300/11852
.................................................................................................... 10400/11852
---
Suite("src/test/codegen") not skipped for "bootstrap::test::Codegen" -- not in ["src/tools/tidy"]
Check compiletest suite=codegen mode=codegen (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 270 tests
ii.....i...i......F..iii........ii.....i.iii.........i.ii.........i.i..............i............i.i. 100/270
..iii........ii.iiii.i..........i.....i...iiii........i..i.i...iii..iiii............................ 200/270
Some tests failed in compiletest suite=codegen mode=codegen host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
.........i.ii...i..i...i....................iiiii.i...................

---- [codegen] codegen/avr/avr-func-addrspace.rs stdout ----

error: compilation failed!
error: compilation failed!
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/codegen/avr/avr-func-addrspace.rs" "-Zthreads=1" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/avr/avr-func-addrspace" "-Crpath" "-Cdebuginfo=0" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-O" "--target=avr-unknown-gnu-atmega328" "--crate-type=rlib" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/avr/avr-func-addrspace/auxiliary" "--emit=llvm-ir"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
error: requires `clone` lang_item

warning: functions generic over types or consts must be mangled
   |
   |
68 |   #[no_mangle]
   |   ------------ help: remove this attribute
69 | / fn call_through_fn_trait(a: &mut impl Fn<(), Output=()>) {
70 | |     (*a)()
   | |_^
   |
   = note: `#[warn(no_mangle_generic_items)]` on by default

---
test result: FAILED. 210 passed; 1 failed; 59 ignored; 0 measured; 0 filtered out; finished in 3.05s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/codegen" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "codegen" "--mode" "codegen" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-10/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "10.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine frontendopenmp fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcerror orcjit passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvutils runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86utils xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap --stage 2 test --exclude src/tools/tidy
Build completed unsuccessfully in 0:13:19
```

</details>

cjgillot commented:
For compiler code, it would be better just to drop the fields. For tests, it may be preferrable to add a dummy `let` just to read the field.

FabianWolff commented:
> For compiler code, it would be better just to drop the fields.

I've dropped the unused fields from the compiler code now, except for the `Error` struct in `compiler/rustc_mir/src/transform/coverage/mod.rs`, where dropping the field amounts to dropping the whole struct:
https://github.com/rust-lang/rust/blob/70e52caed91a43fc01867921e67fcd3478056edd/compiler/rustc_mir/src/transform/coverage/mod.rs#L37-L47
But the struct is used in various places where a `Result` is returned (e.g. [here](https://github.com/rust-lang/rust/blob/70e52caed91a43fc01867921e67fcd3478056edd/compiler/rustc_mir/src/transform/coverage/counters.rs#L47) and [here](https://github.com/rust-lang/rust/blob/70e52caed91a43fc01867921e67fcd3478056edd/compiler/rustc_mir/src/transform/coverage/graph.rs#L347)), so I only left a comment there for now.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2569777944) [(plain)](https://github.com/rust-lang/rust/commit/5077eab8fb72b86dfd7c96db80afb969ee77cae4/checks/2569777944/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
Diff in /checkout/compiler/rustc_mir_build/src/build/matches/simplify.rs at line 172:
                 Ok(())
             }
 
-            PatKind::Binding { name: _, mutability: _, mode, var, ty: _, ref subpattern, is_primary: _ } => {
+            PatKind::Binding {
+                name: _,
+                mutability: _,
+                var,
+                ty: _,
+                ref subpattern,
+                is_primary: _,
+                is_primary: _,
+            } => {
                 candidate.bindings.push(Binding {
                     span: match_pair.pattern.span,
                     source: match_pair.place.clone().into_place(self.tcx, self.typeck_results),
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2018" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_mir_build/src/build/matches/simplify.rs" "/checkout/compiler/rustc_mir_build/src/build/block.rs" "/checkout/compiler/rustc_mir_build/src/build/misc.rs" "/checkout/compiler/rustc_mir_build/src/build/mod.rs" "/checkout/compiler/rustc_mir_build/src/build/scope.rs" "/checkout/compiler/rustc_mir_build/src/build/expr/as_temp.rs" "/checkout/compiler/rustc_mir_build/src/build/expr/as_operand.rs" "/checkout/compiler/rustc_mir_build/src/build/matches/util.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
Build completed unsuccessfully in 0:00:14
```

</details>

matklad commented:
> However, it was pretty straightforward to fix, so did I perhaps overlook something obvious? @matklad, could you weigh in on this?

Looks like I was just mistaken, the impl looks ok to me.

One thing I worry is that, judging by the impact on rustc itself, this would be a noticable user-visible change (ie, big projects are guaranteed to get a couple instances of this at least). So it seems like some design process is needed here, culminating in an FCP. 

I don't know whats the specific appropriate process here is, so let me tag a random t-compiled lead for this: @pnkfelix.

bors commented:
:umbrella: The latest upstream changes (presumably #85328) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #85335) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2592106714) [(plain)](https://github.com/rust-lang/rust/commit/54fc0dbae1f97b4bdf1e7c169dd6b032f6004499/checks/2592106714/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
.................................................................................................... 9500/11870
.................................................................................................... 9600/11870
.........................................................................i......i................... 9700/11870
.................................................................................................... 9800/11870
..................iiiiiii..iiiiii.i................................................................. 9900/11870
.................................................................................................... 10100/11870
.................................................................................................... 10200/11870
.................................................................................................... 10300/11870
.................................................................................................... 10400/11870
---
Suite("src/test/assembly") not skipped for "bootstrap::test::Assembly" -- not in ["src/tools/tidy"]
Check compiletest suite=assembly mode=assembly (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 35 tests
iiiiiiiiiiiiiiiiiiiiiiiiiiiiiii....

 finished in 0.186 seconds
Suite("src/test/incremental") not skipped for "bootstrap::test::Incremental" -- not in ["src/tools/tidy"]
Check compiletest suite=incremental mode=incremental (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
---
Suite("src/test/debuginfo") not skipped for "bootstrap::test::Debuginfo" -- not in ["src/tools/tidy"]
Check compiletest suite=debuginfo mode=debuginfo (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 116 tests
iiiiiiiiii.i.i..i..i..ii....i.i....ii..........iiii.........i.....i...i.......ii.i.ii.....iiii.....i 100/116
test result: ok. 78 passed; 0 failed; 38 ignored; 0 measured; 0 filtered out; finished in 3.81s

 finished in 3.878 seconds
Suite("src/test/ui-fulldeps") not skipped for "bootstrap::test::UiFullDeps" -- not in ["src/tools/tidy"]
---
 finished in 3.169 seconds
Set({"library/std"}) not skipped for "bootstrap::test::Crate" -- not in ["src/tools/tidy"]
Testing std stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
   Compiling std v0.0.0 (/checkout/library/std)
error: field is never read: `flushed`
    |
472 |     pub flushed: bool,
    |     ^^^^^^^^^^^^^^^^^
    |
    |
    = note: `-D dead-code` implied by `-D warnings`
error: aborting due to previous error

error: could not compile `std`


To learn more, run the command again with --verbose.


command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "test" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--locked" "--color" "always" "--features" "panic-unwind backtrace compiler-builtins-c" "--manifest-path" "/checkout/library/test/Cargo.toml" "-p" "std" "--" "--quiet"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap --stage 2 test --exclude src/tools/tidy
Build completed unsuccessfully in 0:20:20
```

</details>

lcnr commented on compiler/rustc_mir/src/transform/coverage/mod.rs:
```diff
@@ -37,6 +37,7 @@ use rustc_span::{CharPos, ExpnKind, Pos, SourceFile, Span, Symbol};
 #[derive(Debug)]
 struct Error {
+    #[allow(dead_code)] // inserted by #85200; maybe remove this whole struct?
     message: String,
```
you can this whole struct here :laughing: cc @richkadel

FabianWolff replied:
I don't understand this comment, sorry!

I didn't remove this struct because it is used in various places as the `Err` component of `Result`, and I wasn't sure what to replace it with (just `()`? An empty `Error` struct? `Option` instead of `Result`?).

lcnr replied:
probably change the functions to be non fallible.

If the function never errors it probably doesn't make much sense to keep returning `Result`

FabianWolff replied:
But the functions _can_ fail, it seems; the error value just never escapes this module:

https://github.com/rust-lang/rust/blob/491cf5561efb1f5ff33c3234ccd0bc5cacbebe3e/compiler/rustc_mir/src/transform/coverage/mod.rs#L265-L267

Perhaps we should keep the `message` field then if it is useful for the `bug!()` output (or just change it to print `e.message` instead of `e`, which would also make the "field never read" warning go away).

richkadel replied:
I don't understand why this attribute is needed. The `coverage::Error` class is used inside the `coverage` module, and `message` is read (AFAICT), and if this somehow needs to be annotated as `dead_code` when it isn't dead code or unread, seems like a hack. It's so trivial that I would assume whatever change might require this would break a lot of other Rust code.

Regarding the comment, I disagree. I don't think it should be changed or dropped. I should be allowed to have a coverage-specific error type (with a message or any other field). This is specifically typed to ensure the `?` try operator applies only to specific coverage function results.

Let me know if I'm missing something.

nikomatsakis replied:
This is an interesting one. The `Debug` is being used here to convey information...hmm.

lcnr commented on compiler/rustc_passes/src/dead.rs:
```diff
@@ -211,6 +211,28 @@ impl<'tcx> MarkSymbolVisitor<'tcx> {
+                            im.of_trait.as_ref().and_then(|impls_trait| impls_trait.trait_def_id());
+                        // Get the DefIds of Clone and Debug
+                        let clone_maybe_def_id =
+                            self.tcx.lang_items().require(hir::LangItem::Clone).ok();
```
```suggestion
                            self.tcx.lang_items().clone_trait();
```

lcnr commented on compiler/rustc_passes/src/dead.rs:
```diff
@@ -220,7 +242,33 @@ impl<'tcx> MarkSymbolVisitor<'tcx> {
+                        .map(|trait_ref| trait_ref.def_id);
+                    // Get the DefIds of Clone and Debug
+                    let clone_maybe_def_id =
+                        self.tcx.lang_items().require(hir::LangItem::Clone).ok();
```
```suggestion
                        self.tcx.lang_items().clone_trait();
```

lcnr commented on compiler/rustc_passes/src/dead.rs:
```diff
@@ -211,6 +211,28 @@ impl<'tcx> MarkSymbolVisitor<'tcx> {
+                        if trait_maybe_def_id.is_some()
+                            && (trait_maybe_def_id == clone_maybe_def_id
+                                || trait_maybe_def_id == debug_maybe_def_id)
+                            && item.span.in_derive_expansion()
```
```suggestion
                            && self.tcx.has_attr(item.def_id, sym::automatically_derived)
```

lcnr commented on compiler/rustc_passes/src/dead.rs:
```diff
@@ -220,7 +242,33 @@ impl<'tcx> MarkSymbolVisitor<'tcx> {
+                    if trait_maybe_def_id.is_some()
+                        && (trait_maybe_def_id == clone_maybe_def_id
+                            || trait_maybe_def_id == debug_maybe_def_id)
+                        && impl_item.span.in_derive_expansion()
```
```suggestion
                        && self.tcx.has_attr(item.def_id, sym::automatically_derived)
```
this feels better to me, it probably doesn't matter too much though.

Would prevent custom derives of `Clone` to profit from this change :thinking: but maybe that's actually a good thing?

FabianWolff replied:
This was a conscious choice: I thought that it might be too confusing to actually see the field being read in your code and still get a warning that the "field is never read". Of course, if you'd prefer, we can ignore all impls of (say) `Clone` and `Debug`, not just automatically derived ones.

lcnr replied:
My edit would not ignore all `Clone` impls but ignore only those with the attribute `#[automatically_derived]` which are added by the compiler for the builtin derives.

This change would cause `Clone` impls emitted by other macros, potentially ones which are better at dealing with implied bounds, to not get ignored by the dead code analysis.

FabianWolff replied:
Ah, I thought you were referring to manual implementations of `Clone`:
```rust
impl Clone for S {
    fn clone(&self) -> Self {
        ...
    }
}
```
I actually included your suggested edit in my last commit; the "conscious choice" I was talking about above referred to manual implementations of `Clone`. Sorry for the confusion!

lcnr commented on library/core/tests/fmt/builders.rs:
```diff
@@ -654,7 +654,9 @@ fn test_formatting_parameters_are_forwarded() {
     use std::collections::{BTreeMap, BTreeSet};
     #[derive(Debug)]
     struct Foo {
+        #[allow(dead_code)]
         bar: u32,
+        #[allow(dead_code)]
         baz: u32,
     }
```
```suggestion
    #[allow(dead_code)]
    struct Foo {
        bar: u32,
        baz: u32,
    }
```

lcnr commented on src/test/ui/closures/2229_closure_analysis/diagnostics/liveness.rs:
```diff
@@ -8,6 +8,7 @@
 struct Point {
     x: i32,
     y: i32,
+    //~^ WARNING: field is never read: `y`
```
i think it still makes sense to add `allow(dead_code)` here as the test cases about something else

lcnr commented on src/test/ui/closures/2229_closure_analysis/diagnostics/liveness_unintentional_copy.rs:
```diff
@@ -6,6 +6,7 @@
 #[derive(Debug)]
 struct MyStruct {
     a: i32,
+    //~^ WARNING: field is never read: `a`
```
same here

lcnr commented on src/test/ui/closures/2229_closure_analysis/run_pass/struct-pattern-matching-with-methods.rs:
```diff
@@ -7,8 +7,13 @@
 #[derive(Debug, Clone, Copy)]
 enum PointType {
     TwoD { x: u32, y: u32 },
+    //~^ WARNING: field is never read: `x`
```
same here

lcnr commented on src/test/ui/closures/2229_closure_analysis/run_pass/unsafe_ptr.rs:
```diff
@@ -9,6 +9,7 @@
 struct S {
     s: String,
     t: String,
+    //~^ WARNING: field is never read: `t`
```
here

lcnr reviewed (commented):
I think this change makes sense. It does have some incorrect warnings if a struct adds a field solely for the sideeffects of that structs `Clone` impl, but that seems like a fine tradeoff to me.

It might also make sense to extend this to all impls with the `syn::automatically_derived` attribute. But the pattern of using fields for their behavior in derived impls is probably more prevalent for traits like `Hash`.

lcnr commented:
nominated @rust-lang/lang signoff

scottmcm commented:
This seems good to me -- the compiler examples of how much this has been hiding is quite strong justification 👍 

If a field is really only for a `Clone` side effect, it getting a `_name` to hide the warning seems totally fine.

One thing I was pondering: What it is about `Clone` and `Debug` specifically that makes it worth excluding them, but not others?  Why wouldn't it make sense to also exclude a derived `PartialEq`, for example?

matklad commented:
>One thing I was pondering: What it is about Clone and Debug specifically that makes it worth excluding them, but not others?

Answer, as usual, history -- it was the two traits that preventing the useful unused warning from firing in the case which prompted me to crate the issue. 

I think, if we are to do this, we should do this for all build-in derives at least.

nikomatsakis commented:
I'm in favor of the concept, I think extending it to all automatically derived traits is pretty logical.

FabianWolff commented:
Thanks for your comments everyone, and thanks for your review and suggestions @lcnr!

I have implemented them now, along with the suggestion to ignore _all_ automatically derived impls, not just those of `Clone` and `Debug`. Predictably, this caused more warnings in the compiler and library code, and as it turned out, many of the fields now (correctly, when ignoring automatically derived impls) marked as "never read" are actually necessary. One example is ordering versions: The fields `major`, `minor`, and `patch` are never explicitly read, but versions are compared, so the fields do have an effect:
https://github.com/rust-lang/rust/blob/a5560a6a90f08a84728802bb5fa5632a17a78672/compiler/rustc_attr/src/builtin.rs#L597-L601

I have prefixed such field names with underscores for now to silence the warning. I think that's acceptable, because the underscore makes explicit that the field is needed only for its "side-effects" (e.g. on ordering).

Also keep in mind that I have already removed most of the actually unnecessary fields in an earlier commit, so the above doesn't mean that this change causes unreasonably many false positives.

richkadel reviewed (commented):
It doesn't make sense (to me) to qualify the `coverage::Error` struct or field. The struct and the `message` field are used. If Rust thinks it isn't used, then something else is wrong with how this is being compiled.

If you must annotate it to workaround another problem, maybe you can add a `FIXME` comment with a bug ID, so this annotation can eventually be removed?

joshtriplett commented:
I do think it's appropriate to exclude Debug and Clone and similar. However, I think PartialEq and PartialOrd are actually legitimate "uses" if and only if the instances are called. It's valid to have a struct with fields that are only used in equality/ordering comparisons, and nowhere else; I think those shouldn't need an underscore.

nikomatsakis commented:
@joshtriplett 

>  I think PartialEq and PartialOrd are actually legitimate "uses" if and only if the instances are called

Agree.

nikomatsakis commented on compiler/rustc_infer/src/infer/region_constraints/mod.rs:
```diff
@@ -269,8 +269,8 @@ pub enum VerifyBound<'tcx> {
 
 #[derive(Copy, Clone, PartialEq, Eq, Hash)]
```
@joshtriplett's point applies to Hash, too.

nikomatsakis reviewed (changes requested):
I hadn't actually looked at the diff before, but I agree that we need to be considering PartialEq, Hash, and friends as using the fields. I think a good goal for the PR should be that "no false warnings" and right now there are a *lot*, but almost all seem to be linked to `PartialEq` / `Eq` / `Hash`. 

The case of the `Error` struct with the `message` field is interesting and a bit more complex.

FabianWolff commented:
So should I go back to just `Clone` and `Debug` for now? Or should I keep the warning for all automatically derived traits except `Eq`, `PartialEq`, `Ord`, `PartialOrd`, and `Hash`?

nikomatsakis commented:
I'm inclined to go back to an "opt-in" scheme, with just `Clone` and `Debug`. I'm trying to put my finger on what makes those traits different. I'd like to articulate the logic behind the heuristic. It seems to be just *more true in practice* that people rarely add fields "just for the purpose of cloning them or having the appear in debug output", although you can certainly imagine the latter (the former is a stretch, in my opinion-- though of course not *impossible*).

That said, this is only a lint. I think that both of those kinds of cases are ones where it'd be reasonable to put a comment explaining what the field is doing there

I think the difference with (e.g.) `PartialEq` is that you could be using the thing a a key in a `BTreeMap`, and therefore making use of those fields in a very intentional way.

FabianWolff committed:
Ignore automatically derived impls of `Clone` and `Debug` in dead code analysis

FabianWolff commented:
> I'm inclined to go back to an "opt-in" scheme, with just `Clone` and `Debug`.

Thanks for the clarification! I have now gone back to just `Clone` and `Debug` again. I've also rebased and squashed my previous commits, so I'd say this is ready for re-review (@lcnr).

nikomatsakis commented:
This is looking quite good to me. I'm going to go ahead and start an FCP merge.

@rfcbot fcp merge

lcnr commented:
I am not able to review any PRs in the near future.

r? @nikomatsakis

bors commented:
:umbrella: The latest upstream changes (presumably #85153) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

cramertj commented:
@nikomatsakis Can you clarify what the final version of the proposal being FCP'd is? It seems like there have been a number of iterations here.

cramertj commented:
@rfcbot concern taylor-isnt-sure-what-they-are-checking-a-box-for

joshtriplett commented:
Could we get explicit confirmation from the proposers here that @rust-lang/lang is only signing off on `Debug` and `Clone` here?

FabianWolff commented:
> Could we get explicit confirmation from the proposers here that @rust-lang/lang is only signing off on `Debug` and `Clone` here?

Yes, this is only about `Debug` and `Clone`.

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2721937816) [(plain)](https://github.com/rust-lang/rust/commit/c11e7e579b44aed669241f1004401bfe81cc6e31/checks/2721937816/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  EXTRA_VARIABLES: {
 "CI_ONLY_WHEN_SUBMODULES_CHANGED": 1
##[endgroup]
adding extra environment variable CI_ONLY_WHEN_SUBMODULES_CHANGED
linux builder detected, using docker to run the build
##[group]Run src/ci/scripts/should-skip-this.sh
---
  IMAGE: x86_64-gnu-tools
##[endgroup]
From https://github.com/rust-lang/rust
 * branch              master     -> FETCH_HEAD
Searching for toolstate changes between 7f9ab0300cd66f6f616e03ea90b2d71af474bf28 and 2236f9267d06075bc23812075a8e79d92cd9de4c
Executing the job since rustdoc was updated
src/ci/scripts/verify-channel.sh
shell: /bin/bash --noprofile --norc -e -o pipefail {0}
env:
  CI_JOB_NAME: x86_64-gnu-tools
---
   |
45 |     path: String,
   |     ^^^^^^^^^^^^
   |
   = note: `-D dead-code` implied by `-D warnings`
error: field is never read: `spans`
  --> src/tools/clippy/clippy_lints/src/regex.rs:54:5
   |
   |
54 |     spans: FxHashSet<Span>,


error: field is never read: `last`
   |
   |
55 |     last: Option<HirId>,

error: aborting due to 3 previous errors

error: could not compile `clippy_lints`
---
   Compiling rls-analysis v0.18.2 (/checkout/src/tools/rls/rls-analysis)
   Compiling futures v0.3.12
   Compiling jsonrpc-derive v17.0.0
   Compiling jsonrpc-core v17.0.0
warning: struct is never constructed: `Signature`
   --> src/tools/rls/rls-analysis/src/analysis.rs:140:12
140 | pub struct Signature {
    |            ^^^^^^^^^
    |
    = note: `#[warn(dead_code)]` on by default
    = note: `#[warn(dead_code)]` on by default

warning: struct is never constructed: `SigElement`
   --> src/tools/rls/rls-analysis/src/analysis.rs:150:12
150 | pub struct SigElement {
    |            ^^^^^^^^^^

   Compiling parity-tokio-ipc v0.8.0
---

warning: field is never read: `spans`
  --> src/tools/clippy/clippy_lints/src/regex.rs:54:5
   |
54 |     spans: FxHashSet<Span>,


warning: field is never read: `last`
   |
   |
55 |     last: Option<HirId>,

   Compiling jsonrpc-core-client v17.0.0
   Compiling rls-ipc v0.1.0 (/checkout/src/tools/rls/rls-ipc)
   Compiling rustc-ap-rustc_arena v718.0.0
---
   |
45 |     path: String,
   |     ^^^^^^^^^^^^
   |
   = note: `-D dead-code` implied by `-D warnings`
error: field is never read: `spans`
  --> src/tools/clippy/clippy_lints/src/regex.rs:54:5
   |
   |
54 |     spans: FxHashSet<Span>,


error: field is never read: `last`
   |
   |
55 |     last: Option<HirId>,

error: aborting due to 3 previous errors

error: could not compile `clippy_lints`
error: could not compile `clippy_lints`

To learn more, run the command again with --verbose.
command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "build" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--locked" "--color" "always" "--manifest-path" "/checkout/src/tools/clippy/Cargo.toml" "--features" "rustc-workspace-hack/all-static" "--message-format" "json-render-diagnostics"
expected success, got: exit status: 101
thread 'main' panicked at 'in-tree tool', src/bootstrap/test.rs:598:14
failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test --stage 2 src/tools/clippy
Build completed unsuccessfully in 0:00:16
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2722561416) [(plain)](https://github.com/rust-lang/rust/commit/bfcba31db652d3b51b3b8fdc0f893b4b56567b52/checks/2722561416/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  EXTRA_VARIABLES: {
 "CI_ONLY_WHEN_SUBMODULES_CHANGED": 1
##[endgroup]
adding extra environment variable CI_ONLY_WHEN_SUBMODULES_CHANGED
linux builder detected, using docker to run the build
##[group]Run src/ci/scripts/should-skip-this.sh
---
  IMAGE: x86_64-gnu-tools
##[endgroup]
From https://github.com/rust-lang/rust
 * branch              master     -> FETCH_HEAD
Searching for toolstate changes between 7f9ab0300cd66f6f616e03ea90b2d71af474bf28 and 6ba2fd335069e4ac1cb12bc738240de8cba5e6c9
Executing the job since clippy or rustfmt subtree was updated
src/ci/scripts/verify-channel.sh
shell: /bin/bash --noprofile --norc -e -o pipefail {0}
env:
  CI_JOB_NAME: x86_64-gnu-tools
---
   Compiling rls-analysis v0.18.2 (/checkout/src/tools/rls/rls-analysis)
   Compiling jsonrpc-core v17.0.0
   Compiling parity-tokio-ipc v0.8.0
   Compiling rustc-ap-rustc_data_structures v718.0.0
warning: struct is never constructed: `Signature`
   --> src/tools/rls/rls-analysis/src/analysis.rs:140:12
140 | pub struct Signature {
    |            ^^^^^^^^^
    |
    = note: `#[warn(dead_code)]` on by default
    = note: `#[warn(dead_code)]` on by default

warning: struct is never constructed: `SigElement`
   --> src/tools/rls/rls-analysis/src/analysis.rs:150:12
150 | pub struct SigElement {
    |            ^^^^^^^^^^

   Compiling jsonrpc-derive v17.0.0
---
---- compile_test stdout ----

error: failed to compile fixed code
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools-bin/clippy-driver" "tests/ui/default_trait_access.fixed" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/test_build_base" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/test_build_base/default_trait_access.stage-id" "--emit=metadata" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/deps" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps" "-Dwarnings" "-Zui-testing" "--extern" "syn=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libsyn-f75d20e88ff41cda.rlib" "--extern" "serde=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libserde-07a2fb86e1f5b8b7.rlib" "--extern" "regex=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libregex-dcbfa95d21c9d3c4.rlib" "--extern" "clippy_lints=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_lints-465fb9c5413ef78c.rlib" "--extern" "quote=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libquote-6083e1acaa6e9ab6.rlib" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/test_build_base/default_trait_access.stage-id.aux"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
{"message":"field is never read: `s`","code":{"code":"dead_code","explanation":null},"level":"error","spans":[{"file_name":"tests/ui/default_trait_access.fixed","byte_start":1707,"byte_end":1720,"line_start":69,"line_end":69,"column_start":5,"column_end":18,"is_primary":true,"text":[{"text":"    pub s: String,","highlight_start":5,"highlight_end":18}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[{"message":"`-D dead-code` implied by `-D warnings`","code":null,"level":"note","spans":[],"children":[],"rendered":null}],"rendered":"error: field is never read: `s`\n  --> tests/ui/default_trait_access.fixed:69:5\n   |\nLL |     pub s: String,\n   |     ^^^^^^^^^^^^^\n   |\n   = note: `-D dead-code` implied by `-D warnings`\n\n"}
{"message":"field is never read: `s`","code":{"code":"dead_code","explanation":null},"level":"error","spans":[{"file_name":"tests/ui/default_trait_access.fixed","byte_start":1816,"byte_end":1824,"line_start":74,"line_end":74,"column_start":5,"column_end":13,"is_primary":true,"text":[{"text":"    pub s: T,","highlight_start":5,"highlight_end":13}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[],"rendered":"error: field is never read: `s`\n  --> tests/ui/default_trait_access.fixed:74:5\n   |\nLL |     pub s: T,\n   |     ^^^^^^^^\n\n"}
{"message":"field is never read: `s`","code":{"code":"dead_code","explanation":null},"level":"error","spans":[{"file_name":"tests/ui/default_trait_access.fixed","byte_start":1888,"byte_end":1911,"line_start":79,"line_end":79,"column_start":5,"column_end":28,"is_primary":true,"text":[{"text":"    pub s: (String, String),","highlight_start":5,"highlight_end":28}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[],"rendered":"error: field is never read: `s`\n  --> tests/ui/default_trait_access.fixed:79:5\n   |\nLL |     pub s: (String, String),\n   |     ^^^^^^^^^^^^^^^^^^^^^^^\n\n"}
{"message":"field is never read: `s`","code":{"code":"dead_code","explanation":null},"level":"error","spans":[{"file_name":"tests/ui/default_trait_access.fixed","byte_start":1975,"byte_end":1994,"line_start":84,"line_end":84,"column_start":5,"column_end":24,"is_primary":true,"text":[{"text":"    pub s: [String; 10],","highlight_start":5,"highlight_end":24}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[],"rendered":"error: field is never read: `s`\n  --> tests/ui/default_trait_access.fixed:84:5\n   |\nLL |     pub s: [String; 10],\n   |     ^^^^^^^^^^^^^^^^^^^\n\n"}

------------------------------------------

thread 'compile_test' panicked at 'Some tests failed', /cargo/registry/src/github.com-1ecc6299db9ec823/compiletest_rs-0.6.0/src/lib.rs:105:22
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2723031499) [(plain)](https://github.com/rust-lang/rust/commit/6c68ac305f842c4314c6946e8d5b04ee137c481d/checks/2723031499/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  EXTRA_VARIABLES: {
 "CI_ONLY_WHEN_SUBMODULES_CHANGED": 1
##[endgroup]
adding extra environment variable CI_ONLY_WHEN_SUBMODULES_CHANGED
linux builder detected, using docker to run the build
##[group]Run src/ci/scripts/should-skip-this.sh
---
  IMAGE: x86_64-gnu-tools
##[endgroup]
From https://github.com/rust-lang/rust
 * branch              master     -> FETCH_HEAD
Searching for toolstate changes between 7f9ab0300cd66f6f616e03ea90b2d71af474bf28 and 0fdf0d99380605fbcc7919fd77b0110d75806c6c
Executing the job since clippy or rustfmt subtree was updated
src/ci/scripts/verify-channel.sh
shell: /bin/bash --noprofile --norc -e -o pipefail {0}
env:
  CI_JOB_NAME: x86_64-gnu-tools
---
   Compiling tokio-util v0.3.1
   Compiling rls-analysis v0.18.2 (/checkout/src/tools/rls/rls-analysis)
   Compiling jsonrpc-core v17.0.0
   Compiling parity-tokio-ipc v0.8.0
warning: struct is never constructed: `Signature`
   --> src/tools/rls/rls-analysis/src/analysis.rs:140:12
140 | pub struct Signature {
    |            ^^^^^^^^^
    |
    = note: `#[warn(dead_code)]` on by default
    = note: `#[warn(dead_code)]` on by default

warning: struct is never constructed: `SigElement`
   --> src/tools/rls/rls-analysis/src/analysis.rs:150:12
150 | pub struct SigElement {
    |            ^^^^^^^^^^

   Compiling jsonrpc-derive v17.0.0
---
    |
768 |     span: Span,
    |     ^^^^^^^^^^
    |
    = note: `-D dead-code` implied by `-D warnings`
error: field is never read: `attrs`
  --> src/tools/rustfmt/src/modules.rs:30:5
   |
   |
30 |     attrs: Cow<'a, Vec<ast::Attribute>>,

error: aborting due to 2 previous errors

error: could not compile `rustfmt-nightly`
error: could not compile `rustfmt-nightly`

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: build failed
command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "build" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--locked" "--color" "always" "--manifest-path" "/checkout/src/tools/rustfmt/Cargo.toml" "--features" "rustc-workspace-hack/all-static" "--message-format" "json-render-diagnostics"
expected success, got: exit status: 101
thread 'main' panicked at 'in-tree tool', src/bootstrap/test.rs:324:14
failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test --stage 2 src/tools/rustfmt
Build completed unsuccessfully in 0:00:14
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #85904) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

nikomatsakis commented:
@cramertj 

> rfcbot concern taylor-isnt-sure-what-they-are-checking-a-box-for

Do you feel this concern is resolved? :)

cramertj commented:
@nikomatsakis I think so-- the plan is to start linting on fields that are unused except in `Clone` or `Debug` derives, but not other derives. I'm curious if there's a general principle we plan to apply here, or if we're thinking of this as a one-off "Clone and Debug are especially uninteresting". For example, I could also imagine wanting a lint on a `StructOpt` field that was never used, as that could indicate that the commandline option was not respected or should be removed. That said, I think I'm fine signing off on this, so..

@rfcbot resolve taylor-isnt-sure-what-they-are-checking-a-box-for

Xiphoseer commented:
To me, the difference between `Clone` and `Debug` vs `{Partial,}Eq`, `{Partial,}Ord` and `Hash`, is whether there's data actually flowing out of them. `Clone` is logically the identity function, so it doesn't add anything to the set of meaningful use sites and Debug has no defined output format, so it's only ever useful as a printout to a person. Which is probably why excluding them adds little false-positives.

The only other trait method I can think of that has that kind of "no meaningful return value" semantics and could potentially access all field is `Iterator::size_hint` (without `TrustedLen`) – I think it'd be nice to be warned about fields that are used in `size_hint` but not in `next`. But that's not a derivable trait. `{Partial,}Eq`, `{Partial,}Ord` or `Hash` on a unit struct would be another case.

<details>
<summary>Aside: StructOpt doesn't actually seem to prevent dead code detection in any way</summary>

```rust
use structopt::StructOpt;

#[derive(StructOpt)]
pub struct Test {
  foo: u32,
  bar: u32,
}

fn main() {
    let a = Test::from_args();

    println!("Hello, {}!", a.bar);
}
```
</details>

Havvy commented:
I don't like the fact that this is _not_ a visible change in the standard library. I'd like to see an attribute on these derived impls `#[trivial_field_reads]` (*bikesheddable*). So, for example, the macro expansion becomes:

```
#![feature(prelude_import)]
#[prelude_import]
use std::prelude::rust_2018::*;
#[macro_use]
extern crate std;
struct Struct {
    field: i32,
}
#[automatically_derived]
#[trivial_field_reads]
#[allow(unused_qualifications)]
impl ::core::clone::Clone for Struct {
    #[inline]
    fn clone(&self) -> Struct {
        match *self {
            Struct { field: ref __self_0_0 } =>
            Struct{field: ::core::clone::Clone::clone(&(*__self_0_0)),},
        }
    }
}
```

bors commented:
:umbrella: The latest upstream changes (presumably #86627) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #86704) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #86841) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3209082407) [(plain)](https://github.com/rust-lang/rust/commit/52012b3aa9029f01eab3d766a193a56f7a72cb8c/checks/3209082407/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   |
68 |     parent: HirId,
   |     ^^^^^^^^^^^^^
   |
   = note: `-D dead-code` implied by `-D warnings`

error: field is never read: `node`
  --> compiler/rustc_middle/src/hir/map/collector.rs:69:5
   |
69 |     node: Node<'hir>,

error: aborting due to 2 previous errors

error: could not compile `rustc_middle`
```

</details>

nikomatsakis commented on compiler/rustc_passes/src/dead.rs:
```diff
@@ -242,6 +242,24 @@ impl<'tcx> MarkSymbolVisitor<'tcx> {
         }
     }
 
+    fn should_ignore_impl(&self, impl_def_id: DefId) -> bool {
```
Can we add a doc comment explaining the purpose of this function and why those particular traits were chosen? (Perhaps linking to this PR)

FabianWolff replied:
Fixed, along with your other suggestion.

nikomatsakis commented on compiler/rustc_passes/src/dead.rs:
```diff
@@ -266,6 +284,17 @@ impl<'tcx> MarkSymbolVisitor<'tcx> {
+                            // automatically derived via #[derive(...)]. Ignore this for dead code
+                            // analysis (derived implementations of Debug and Clone always trivially
+                            // use all fields, so they would hide unused field warnings for such
+                            // structs/enums, cf. issue #84647).
```
Maybe just move this comment and change this code to `if !self.should_ignore_impl(...) { .. }`

nikomatsakis reviewed (changes requested):
r=me after comments are tweaked

nikomatsakis commented:
I do tend to agree with @Havvy that it'd be nice to have the code use a `#[rustc_foo]` attribute  to decide which traits to ignore, but I'm not going to block on that. @FabianWolff if you want I can explain how to do that.

FabianWolff commented:
> I do tend to agree with @Havvy that it'd be nice to have the code use a `#[rustc_foo]` attribute to decide which traits to ignore, but I'm not going to block on that. @FabianWolff if you want I can explain how to do that.

Thanks for your review @nikomatsakis! I have implemented @Havvy's `rustc_trivial_field_reads` attribute, so the change is now visible when looking at the macro expansions, e.g.
```rust
#[derive(Debug)]
struct S { i: i32 }
```
becomes
```rust
[...]
#[automatically_derived]
#[allow(unused_qualifications)]
impl ::core::fmt::Debug for S {
    #[rustc_trivial_field_reads]
    fn fmt(&self, f: &mut ::core::fmt::Formatter) -> ::core::fmt::Result {
        [...]
    }
}
```
The change is still not visible "in the standard library" as suggested by https://github.com/rust-lang/rust/pull/85200#issuecomment-859980356, though, because derive macro expansion happens in the `rustc_builtin_macros` part of the compiler.

bors commented:
:umbrella: The latest upstream changes (presumably #88039) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nikomatsakis commented on compiler/rustc_feature/src/builtin_attrs.rs:
```diff
@@ -458,6 +458,8 @@ pub const BUILTIN_ATTRIBUTES: &[BuiltinAttribute] = &[
     // Enumerates "identity-like" conversion methods to suggest on type mismatch.
     rustc_attr!(rustc_conversion_suggestion, AssumedUsed, template!(Word), INTERNAL_UNSTABLE),
+    // Causes field reads in the annotated item to be ignored during dead code analysis.
+    ungated!(rustc_trivial_field_reads, AssumedUsed, template!(Word)),
```
wait, why is this `ungated!`? that doesn't seem right

nikomatsakis replied:
doesn't that mean anybody can use it? that's not what we want :)

FabianWolff replied:
I initially thought so, too, but remember that this is generated in user code as part of the derive macro expansion, so anybody _has_ to be able to use it in order to use the derive macro (in my initial implementation, I gated the attribute and received lots of errors about uses of unstable/internal features in the expanded user code). Or is there a more elegant way of resolving this problem?

nikomatsakis replied:
Well, we have mechanisms for derives and things to use unstable code, but an alternative would be to put the attribute on the **trait**, rather than the impl. We would then detect an auto-generated impl for a trait with that attribute. (Plausibly this could apply to hand-written code, too...)

nikomatsakis replied:
I think I would prefer the attribute on the trait, I think it'd be more visible, and it's more obvious how to extend to other traits should we decide to do so in the future.

nikomatsakis replied:
Thoughts, @FabianWolff ?

FabianWolff replied:
@nikomatsakis Done. Sorry for taking so long to implement this.

I have decided not to extend this to manual implementations, though, i.e. my latest changes still only apply to automatically derived `impl`s. I think this would otherwise become too confusing/annoying; for instance, when ignoring all `Clone` implementations, including, say, this one:
https://github.com/rust-lang/rust/blob/69c4aa2901ffadf69deaf91b2f90604bcbc2eb36/library/core/src/ffi.rs#L364-L371

the `va_copy()` function would be reported as dead code, but we can't actually delete it, because it _is_ used. The same situation can arise for field reads if someone chooses to manually implement `Clone`/`Debug` for whatever reason.

sukhmel replied:
in our codebase, we've recently discovered that some of our `proc_macro_derive`s have been hiding dead code. The derive uses fields to create an implementation for a different structure but fields must be accessed in order for that to have any use.

Am I correct in understanding that after `rustc_trivial_field_reads` has been gated, there was no trait or a different way to mark our derive to be ignored during dead code analysis? I only see something similar in shepmaster's [comment](https://github.com/rust-lang/rust/pull/85200#issuecomment-1010371394) lower, and that is never addressed. Also, I'm a bit confused as to where to ask about this problem, if I should open an issue somewhere, or if there is some workaround. 

Ironically, everything I was able to google up was talking about how to silence the warning, not the other way around.

sukhmel replied:
Update regarding my question, we've managed to redo derive a bit so that it does not read fields and restrict visibility to  get dead_code warnings back

bors commented:
:umbrella: The latest upstream changes (presumably #87739) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #80522) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nikomatsakis reviewed (approved):
This looks great!

bors commented:
:v: @FabianWolff can now approve this pull request
<!-- homu: {"type":"Delegated","delegator":"nikomatsakis","delegate":"FabianWolff"} -->

FabianWolff commented:
@bors r=nikomatsakis

bors commented:
:pushpin: Commit 79adda930f9b607ecb4819ed7abcf1cd285e938a has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 79adda930f9b607ecb4819ed7abcf1cd285e938a -->
<!-- homu: {"type":"Approved","sha":"79adda930f9b607ecb4819ed7abcf1cd285e938a","approver":"nikomatsakis"} -->

bors commented:
:evergreen_tree: The tree is currently [closed](https://github.com/rust-lang/rust/pull/88783#issuecomment-916338297) for pull requests below priority 100. This pull request will be tested once the tree is reopened.

calebcartwright commented:
It'd be most helpful and appreciated in the future if folks would consider cc'ing me or the rustfmt wg when any non-trivial changes are being made to rustfmt, if for no reason other than to help me keep up with changes being made to rustfmt outside r-l/rustfmt. Unfortunately the High Five bot pings only work when there's subtree modifications at the time the PR is opened, so we miss any and all updates that are applied to subtree tools after the initial PR.

Manishearth commented:
This may have impacted perf: https://github.com/rust-lang/rust/pull/88824#issuecomment-932664761

saethlin commented:
Can this be added to the 1.57 release notes? It's a pretty substantial change, because of how common `#![deny(dead_code)]` is.

shepmaster commented:
> I'd like to see an attribute on these derived impls

> it'd be nice to have the code use a `#[rustc_foo]` attribute

Is there another issue I can follow to track the stabilization	of this attribute? One of my crates uses derives to create structures and those structures effectively disable the dead code lint. I believe this attribute might help me in my case.
